File: 3rd/yitong-ren-redux.py
# https://towardsdatascience.com/a-step-by-step-implementation-of-gradient-descent-and-backpropagation-d58bda486110
import numpy as np
class NeuralNetwork:
    def __init__(self):
        np.random.seed(10) # for generating the same results
        self.weights = [0,0]
        self.weights[0] = np.random.rand(3,4) # input to hidden layer weights
        self.weights[1] = np.random.rand(4,1) # hidden layer to output weights
        
    def sigmoid(self, z):
        return 1/(1 + np.exp(-z))
    
    def sigmoid_derivative(self, z):
        return self.sigmoid(z) * (1 - self.sigmoid(z))
    
    def gradient_descent(self, x, y, iterations):
        layers = [0,0,0]
        for i in range(iterations):
            ##############################################################################
            # forward

            layers[0] = x
            layers[1] = self.sigmoid( np.dot( layers[0], self.weights[0] ) )
            layers[2] = self.sigmoid( np.dot( layers[1], self.weights[1] ) )

            ##############################################################################
            # back propagation

            gradients = [0,0] 
            delta = [0,0] 
            lastWeightIndex = len( self.weights ) - 1
            for i in range( lastWeightIndex, -1, -1 ):
                slope = self.sigmoid_derivative( np.dot( layers[i], self.weights[i] ) )
                if i == lastWeightIndex:
                    error = y - layers[2]
                else:
                    error = np.dot(delta[i+1], self.weights[i+1].T)
                delta[i] = error * slope
                gradients[i] = np.dot(layers[i].T, delta[i]) 

            # update weights
            for i in range( len( self.weights ) ):
                self.weights[i] += gradients[i]

        print('The final prediction from neural network are: ')
        print(layers[2])
if __name__ == '__main__':
    neural_network = NeuralNetwork()
    print('Random starting input to hidden weights: ')
    print(neural_network.weights[0])
    print('Random starting hidden to output weights: ')
    print(neural_network.weights[1])
    X = np.array([[0, 0, 1], [1, 1, 1], [1, 0, 1], [0, 1, 1]])
    y = np.array([[0, 1, 1, 0]]).T
    neural_network.gradient_descent(X, y, 10000)

File: 3rd/yitong-ren-redux.js
#!/usr/bin/env node 

// adapted from https://towardsdatascience.com/a-step-by-step-implementation-of-gradient-descent-and-backpropagation-d58bda486110
const nj = require('numjs');
class NeuralNetwork {
	constructor() {
        // nj.random.seed(10) // for generating the same results
        this.weights = [
        	  nj.random(3,4) // input to hidden layer weights
        	, nj.random(4,1) // hidden layer to output weights
		];
	}

    sigmoid( z ) {
		if ( !isNaN( z ) ) {
			return 1 / (1 + Math.exp(-z) );
		}
		const a = z.selection.data;
		a.forEach( (v,i) => a[i] = this.sigmoid( v ) );
		return z;
	}

    sigmoid_derivative(z) {
		if ( !isNaN( z ) ) {
			const q = this.sigmoid(z) 
			return q * ( 1 - q );
		}

		const a = z.selection.data;
		a.forEach( (v,i) => a[ i ] = this.sigmoid_derivative( v ) );
		return z;
	}

	print( message, a ) {
		console.log( message + '\n' + JSON.parse( a.toJSON() ).map(r=>`\t[${r.join(', ')}]`).join('\n')  );
	}
    
    gradient_descent( x, y, iterations ) {
        const lastWeightIndex = this.weights.length - 1;
		let result = null;

		for ( let t = 0 ; t < iterations ; t++ ) {
            //////////////////////////////////////////////////////////////////////////////
            // forward propagation

			const layers = new Array( this.weights.length + 1 );
			for ( let i = 0 ; i < layers.length ; i++ ) {
				const j = i - 1;
				result = layers[ i ] = i ? this.sigmoid( nj.dot( layers[j], this.weights[j] ) ) : x;
			}

            //////////////////////////////////////////////////////////////////////////////
            // back propagation

            const gradients = new Array( this.weights.length );
			const delta = new Array( this.weights.length );

			for ( let i = lastWeightIndex ; i > -1 ; i-- ) {
                const slope = this.sigmoid_derivative( nj.dot( layers[i], this.weights[i] ) );
				const error = ( i == lastWeightIndex
					? nj.subtract( y , layers[i+1] )
                    : nj.dot(delta[i+1], this.weights[i+1].T)
				);

                delta[i] = nj.multiply( error , slope );
                gradients[i] = nj.dot( layers[i].T, delta[i] );
			}

            // update weights
			for ( let i = 0 ; i <= lastWeightIndex ; i++ ) {
                this.weights[i] = nj.add( this.weights[i], gradients[i] )
			}
		}

		this.print( 'The final prediction from neural network are:', result );
	}
}

const main = () => {
    const neural_network = new NeuralNetwork()
	neural_network.print( 
    	'Random starting input to hidden weights: ',
    	neural_network.weights[0]
	)
	neural_network.print( 
    	'Random starting hidden to output weights: ',
    	neural_network.weights[1]
	);

    X = nj.array([[0, 0, 1], [1, 1, 1], [1, 0, 1], [0, 1, 1]])
    y = nj.array([[0, 1, 1, 0]]).T
    neural_network.gradient_descent(X, y, 10000)
}

main();
